### PR TITLE
CBMC proofs for aws_hash_table_clear() and aws_hash_table_clean_up()

### PR DESCRIPTION
**Pull request by danielsn**
*Description of changes:*
* aws_hash_table_clear CBMC proof 
* aws_hash_table_clear code cleanup.
* aws_hash_table_clean_up CBMC proof
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .cbmc-batch/jobs/aws_hash_table_clear/aws_hash_table_clear_harness.c [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+#include <proof_helpers/proof_allocators.h>
+#include <proof_helpers/utils.h>
+
+void aws_hash_table_clear_harness() {
+
+    struct aws_hash_table map;
+    ensure_allocated_hash_table(&map, MAX_TABLE_SIZE);
+    __CPROVER_assume(aws_hash_table_is_valid(&map));
+    map.p_impl->destroy_key_fn = nondet_bool() ? NULL : hash_proof_destroy_noop;
+    map.p_impl->destroy_value_fn = nondet_bool() ? NULL : hash_proof_destroy_noop;
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_hash_table_clear/aws_hash_table_clear_harness.c]: Same comment here to move this to a separate function.

File: source/hash_table.c [hunk 2]
```diff
@@ -221,6 +221,8 @@ int aws_hash_table_init(
 }
 
 void aws_hash_table_clean_up(struct aws_hash_table *map) {
+    AWS_PRECONDITION(map);
+    AWS_PRECONDITION(map->p_impl == NULL || aws_hash_table_is_valid(map));
```
Comment (feliperodri) [inline source/hash_table.c]: `map->p_impl->alloc. != NULL` ?

File: source/hash_table.c [hunk 3]
```diff
@@ -229,9 +231,10 @@ void aws_hash_table_clean_up(struct aws_hash_table *map) {
     }
 
     aws_hash_table_clear(map);
-    aws_mem_release(((struct hash_table_state *)map->p_impl)->alloc, map->p_impl);
+    aws_mem_release(map->p_impl->alloc, map->p_impl);
 
     map->p_impl = NULL;
+    AWS_PRECONDITION(map->p_impl == NULL);
```
Comment (feliperodri) [inline source/hash_table.c]: `AWS_POSTCONDITION`

File: .cbmc-batch/jobs/aws_hash_table_clean_up/Makefile [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+# or in the "license" file accompanying this file. This file is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied. See the License for the specific language governing permissions and
+# limitations under the License.
+
+###########
+#4: 24s
+#8: 50s
+#16: 6m30s
+MAX_TABLE_SIZE ?= 8
```
[the remark this answers is not in the document]
Comment (danielsn) [reply .cbmc-batch/jobs/aws_hash_table_clean_up/Makefile]: Done

File: .cbmc-batch/jobs/aws_hash_table_clean_up/Makefile [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+DEFINES += -DMAX_TABLE_SIZE=$(MAX_TABLE_SIZE) 
+
+UNWINDSET += aws_hash_table_clear.0:$(shell echo $$((1 + $(MAX_TABLE_SIZE))))
+UNWINDSET += memset_override_0_impl.0:$(shell echo $$(($$((3 * $(MAX_TABLE_SIZE))) + 11)))
```
[the remark this answers is not in the document]
Comment (danielsn) [reply .cbmc-batch/jobs/aws_hash_table_clean_up/Makefile]: Done

File: .cbmc-batch/jobs/aws_hash_table_clean_up/aws_hash_table_clean_up_harness.c [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+#include <proof_helpers/utils.h>
+
+void aws_hash_table_clean_up_harness() {
+
+    struct aws_hash_table map;
+    ensure_allocated_hash_table(&map, MAX_TABLE_SIZE);
+    __CPROVER_assume(aws_hash_table_is_valid(&map));
+    map.p_impl->destroy_key_fn = nondet_bool() ? NULL : hash_proof_destroy_noop;
+    map.p_impl->destroy_value_fn = nondet_bool() ? NULL : hash_proof_destroy_noop;
+    map.p_impl->alloc = can_fail_allocator();
```
[the remark this answers is not in the document]
Comment (danielsn) [reply .cbmc-batch/jobs/aws_hash_table_clean_up/aws_hash_table_clean_up_harness.c]: That's already checked in hash_table_is_valid()

File: .cbmc-batch/jobs/aws_hash_table_clean_up/aws_hash_table_clean_up_harness.c [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+#include <proof_helpers/proof_allocators.h>
+#include <proof_helpers/utils.h>
+
+void aws_hash_table_clean_up_harness() {
+
+    struct aws_hash_table map;
+    ensure_allocated_hash_table(&map, MAX_TABLE_SIZE);
+    __CPROVER_assume(aws_hash_table_is_valid(&map));
+    map.p_impl->destroy_key_fn = nondet_bool() ? NULL : hash_proof_destroy_noop;
+    map.p_impl->destroy_value_fn = nondet_bool() ? NULL : hash_proof_destroy_noop;
```
[the remark this answers is not in the document]
Comment (danielsn) [reply .cbmc-batch/jobs/aws_hash_table_clean_up/aws_hash_table_clean_up_harness.c]: Done

File: .cbmc-batch/jobs/aws_hash_table_clean_up/aws_hash_table_clean_up_harness.c [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    map.p_impl->alloc = can_fail_allocator();
+
+    size_t empty_slot_idx;
+    __CPROVER_assume(aws_hash_table_has_an_empty_slot(&map, &empty_slot_idx));
+    aws_hash_table_clean_up(&map);
+    assert(map.p_impl == NULL);
```
[the remark this answers is not in the document]
Comment (danielsn) [reply .cbmc-batch/jobs/aws_hash_table_clean_up/aws_hash_table_clean_up_harness.c]: The difficulty is that the bytes have already been freed at this point.  But I can use the same trick I did for strings and keep it allocated.
Done

File: .cbmc-batch/jobs/aws_hash_table_clear/Makefile [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+###########
+#4: 35s
+#8: 1m10s
+#16: 6m30s
+MAX_TABLE_SIZE ?= 8
+DEFINES += -DMAX_TABLE_SIZE=$(MAX_TABLE_SIZE) 
+
+UNWINDSET += aws_hash_table_clear.0:$(shell echo $$((1 + $(MAX_TABLE_SIZE))))
+UNWINDSET += memset_override_0_impl.0:$(shell echo $$(($$((3 * $(MAX_TABLE_SIZE))) + 11)))
```
[the remark this answers is not in the document]
Comment (danielsn) [reply .cbmc-batch/jobs/aws_hash_table_clear/Makefile]: Done

File: .cbmc-batch/jobs/aws_hash_table_clear/aws_hash_table_clear_harness.c (hunk 1 above)
Comment (danielsn) [reply .cbmc-batch/jobs/aws_hash_table_clear/aws_hash_table_clear_harness.c]: Done

File: source/hash_table.c (hunk 2 above)
Comment (danielsn) [reply source/hash_table.c]: included in hash_table_is_valid

File: source/hash_table.c (hunk 3 above)
Comment (danielsn) [reply source/hash_table.c]: Fixed

**Pull request merged**